### DISCUSSION TITLE
How to flash Tasmota on an ESP8285-based Elivco Tuya door sensor

### DISCUSSION THREAD

**Discussion by spupetic** [arendst/Tasmota#20559] [Support request]
Update: The solution is to put a solder blob or a little wire to the place of the missing R2, then you can back up the original firmware and flash it with Tasmota. **Before you do that, please sniff the communication between JS2303 and the ESP8285 chip while having the Tuya firmware on the device.**

Original question:

Hi everyone! I would like to ask for some advice about flashing a generic **Elivco** Tuya door sensor with Tasmota, as I have tried to do it via serial and also with tuya-convert, but I could not do it. The pins are exposed and labeled, I have connected it properly, but it seems like the RX on the board is not active somehow, the TX line sends data (Tuya logs when not in flash mode), but somewhy I cannot get it to display correctly (tried several baud rates). esptool times out, I cannot read the flash, the web flasher does not work either. I have tried to ground GPIO0 to enter flash mode, that seems successful, but from there, I cannot get it to flash. 

![tuya_door_02](https://github.com/arendst/Tasmota/assets/144851235/e44de9a4-9d4a-4a6f-bc06-61b3b244164c)
![tuya_door_01](https://github.com/arendst/Tasmota/assets/144851235/7b8e8a3f-5ad3-4fea-8a1d-ddadd72249ce)

I have also tried powering the device externally, that did not help.

The label on the board says: `FL-S157-V1.3`

**Comment by barbudor** [comment]
Your device doesn't seems to have a Tuya MCU which is generally the cause of such trouble as it drives the RX pin of the ESP.
If you are sure that there's nothing connected to the ESP's RX pins, the other pain point on battery operated device is a circuit that resets the ESP to force a low power mode. Check what could be connected to the RST or the CH_EN pins

**Reply by sfromis** [reply]
You could also try keeping the sensor active while working with it, giving it no rest, as a door sensor may indeed power off the ESP8266 shortly after last detected movement.

**Comment by spupetic** [comment]
Thank you for the tips, I have tried again, this time, moving the magnet around, and it started flashing, but only got to the point where it erased the flash, so now the original firmware is gone. It seems like that the hall sensor powers on the ESP, but with the original firmware, it went along and created an AP without the circuit shutting it down. What can I do, to keep it powered on? There are some other pads exposed, I suspect, that there should be a mode what allows flashing.

**Reply by sfromis** [reply]
Speculation: Maybe the 6-pin chip near the center hole is some sort of timer, it does seem to have connections to the ESP8285 chip, which might be able to control the timing. In that case, it might be feasible to supply power around that chip. Beware that, in that case, you might have to work out how to manage the timing from Tasmota.

**Reply by spupetic** [reply]
Thanks, I will check it out.

**Comment by CoderMal** [comment]
I have the same unit. I was able to flash it using Tasmotizer but cannot figure out how to control the JC2303.  Does anyone have a datasheet for the JC2303?  Searching comes up empty. It is used to control the CHIP_EN through GPIO4 and is tied to both GPIO5 (hall sensor) and GPIO12 (button on bottom) for wakeup. 

JC2303 Pinout is:
1 - GPIO12
2 - GND
3 - GPIO5
4 - CHIP_EN
5 - VCC
6 - GPIO4

**Reply by spupetic** [reply]
How did you flash it with Tasmotizer?

**Reply by CoderMal** [reply]
If you already configured the device via a Tuya app the device will quickly go to sleep before you get a chance to flash it. You will need to reset the device by holding the reset button on the bottom until the led blinks quickly.

Before flashing. Please check how the ESP controls the JC2303 with an oscilloscope or logic analyser.  I cannot find the datasheet and without knowing the communications, you will not be able to put the ESP into power-down state, thus killing batteries in a few days. 

Please post your findings.

**Reply by spupetic** [reply]
Unfortunately, I have already erased it while trying to flash, it only went through that point via esptool, and cannot even started with Tasmota's web flasher, my unit is now bricked. I don't have the Tuya firmware anymore, but the chip still wakes up the ESP when moving the magnet around.

**Reply by CoderMal** [reply]
Have you tried to putting CHIP_EN to VCC (bypassing the JS2303)? You can do this with a little wire or solder blob on the missing R2 by the CHIP_EN pin.

Tasmotizer running on windows has always been my go-to. https://github.com/tasmota/tasmotizer

**Reply by spupetic** [reply]
Unfortunately, I did not have time to check it. I have tried to look up the JS2303 online too, had no luck with it. But... Do we need it? Cannot we just put the ESP into deep sleep after it sent the MQTT message? The ESP wakes up to a door event without firmware on it, so the controls may not require other communication, like the JS2303 resets the ESP, it wakes up, sends the message, and the software turns it to deep sleep.

**Comment by spupetic** [comment]
> Have you tried to putting CHIP_EN to VCC (bypassing the JS2303)? You can do this with a little wire or solder blob on the missing R2 by the CHIP_EN pin.
> 
> Tasmotizer running on windows has always been my go-to. https://github.com/tasmota/tasmotizer

Soldering onto the missing R2 worked, thank you. Now the device runs Tasmota. Have you been able to figure out what can be done about the JS2303?

**Reply by CoderMal** [reply]
Glad it worked for you. I think the JS2303 on my board is damaged. I order another unit so when it arrives, I will confirm.

**Comment by spupetic** [comment]
> Glad it worked for you. I think the JS2303 on my board is damaged. I order another unit so when it arrives, I will confirm.

I see. What could have happened to it?

**Comment by burbon-pl** [comment]
Hello,

Did anyone succeed to make this thing running? I've managed to upload tasmota (R2 shorted), found out that LED is probably on GPIO13, TX at GPIO1,RX GPIO3. Hall sensor imho should be on GPIO4, as there is the JS2303 connected.
JS2303 wakes this thing up for so short time I don't think it's possible to connect to wifi and send anything. Anyone tried bypassing JS2303 (connecting hall sensor directly to ESP8285, and shortening R2)?

**Comment by theanch92** [comment]
Hello,

I don't know if anyone is still interested, but I've done a bit of reverse engineering.
The JS2303 chip reads the magnetic sensor on Pin 3 (MOD - GPIO5 on ESP). When its state changes (Low for closed door and High for open door), the Pin 4 (RSTB - CHIP_EN) pin is enabled, turning on the ESP. It then waits for the ESP to set the SCK - GPIO4 pin to a HIGH state; if this is not done, it turns off the CHIP_EN. In this state, the ESP operates normally (pairing the device or sending the alarm by reading the status of GPIO5), then it sets GPIO4 to low, and the JS2303 turns off the CHIP_EN. Pin SDA-GPIO12 appears to be unused.

<img width="1027" height="361" alt="Screenshot 2025-10-05 153952" src="https://github.com/user-attachments/assets/2560b051-0cd4-4efa-ace0-c7a5e69f92a1" />

With Tasmota this is my configuration:

GPIO4: Relay1
GPIO5: Switch1
PowerOnState 1  **(Essential, otherwise the ESP is turned off immediately)**
SwitchMode 1
SetOption114 1
SetOption36 0
SetOption65 1
Rule1 ON Mqtt#Connected DO RuleTimer1 3 ENDON ON Rules#Timer=1 DO Power1 0 ENDON **(Send data and then sends the power-off command)**

I'm not very familiar with tasmota but with these settings it seems to work correctly for me.

**Comment by aunt-lydi4** [comment]
Hello and thank you for this post. It helped me in my long fight with this sensor, especially in the process of understanding how it works. I had a box that was sitting in my shelve somewhere full of Tuya devices which needed to be flashed with open firmware, and literally this was the last one and it was the boss. To quickly get to the gist: After doing some tinkering, the JC2301 (U4) seemed to have stopped working, so I had to find a way to make it work without it, which I want to share here.

My board revision is "V1.2", which is probably one reason why the JC unit has another part number, however I assume that they work in the same way. I paste my notes here and then some photos to explain how I solved the problem.

	- Device wake-up / low power logic was controlled by IC "JC2301"
	- This unit switched CHIP_EN on and off based on sensor HIGH/LOW event and reports from ESP GPIO4 (P16)*
	- The IC interrupted flashing process due to controlling ESP's power state
	- The IC was removed for testing purposes during attempts to flash Tasmota, it misbehaved and got either physically or electronically damaged during tests/removal/reinstallation
	- The chip's logic enabled ultra-low power consumption as it completely shut off ESP between trigger events
	- A replacement circuit ("Edge-to-Reset Converter") was necessary to rebuild low-power operation of the module
	- This circuit takes advantage of ESP's deep sleep ability:
		○ Deep sleep can be enabled through firmware command (DeepSleepTime xx)
		○ Only the set timer value will wake up ESP through EXT_RSTB (P32) via XDP_DCDC/GPIO16 (P8)**
		○ Pulling EXT_RSTB to GND (LOW) also triggers a wakeup (normal "Reset")
	- So the idea is to generate pulses to pull RST low upon sensor events (magnet approach/removal)


<img width="751" height="343" alt="Screenshot 2026-04-12 180014" src="https://github.com/user-attachments/assets/debcc4b8-caf4-496c-9b49-de63159f0803" />

	- The circuit above was tested with following power supplies: bench PSU, low voltage regulator and 2x 1.2V rechargeable batteries (Type AAA)
	- It was tested with standard tasmota.bin firmware under normal operations and deepsleep operation and confirmed working: every magnet event (removal and approach) triggers RST, so the ESP reboots and then goes back to deep sleep (if deepsleep is enabled)
	- The trigger on magnet approach is achieved through 24nF ceramic capacitor between RST and Hall OUT
	- The trigger on magnet removal is achieved through NPN transistor (BC547) wired to RST (Collector), Hall-OUT through 11nF capacitor (Base) and GND (Emitter) with 330k resistor across Base and Emitter
	- An optional 24nF capacitor was suggested to be placed across RST and GND to improve stability and reduce noise, however after assembly and during operation it seems to have disturbed the converter instead of improving it, so it was removed again
	- although during the first few attempts during normal operation in smart home, the trigger did not work reliably 
	- After few days operation under normal daily conditions (opening a window for 20 min and then close it, around 2 times a day), it works without problems, the reasons for unstable and now stable operation is not known
	- supposedly the pulses were not strong enough to trigger RST to wake up during deep sleep, however in always-on mode without deepsleep, it worked instantly and flawlessly

*as already mentioned earlier in this post, a pull-up resistor (I believe it was 10k) to power CHIP_EN was intentionally omitted by the manufacturer, so I added it in the place that was already prepared for it on the board
**on this board revision V1.2, a trace cut was already placed by the manufacturer on the connection from EXT_RSTB to XDP_DCDC; presumably to prevent any interference with the power management of JC2301. I restored the connection to have deep sleep timer function, even though it wouldn't be necessary as the chip awakes from the edge circuit, but it's a nice feature to have)


Here some photos for reference, relevant pins, solder points, and my little extra module:

![20260406_022130](https://github.com/user-attachments/assets/09de78ab-a2b1-4d07-b43a-c22350b5a82f)
![20260405_032725](https://github.com/user-attachments/assets/3806f82d-e361-4148-9a67-b7b30714871f)
![20260406_020904_edited](https://github.com/user-attachments/assets/32d4b37c-4ae8-440f-84a4-753c8fa23d26)
![20260406_021801](https://github.com/user-attachments/assets/6cacbb1f-6217-49a3-af72-9c30e2c3a20c)
![20260405_220805_edited](https://github.com/user-attachments/assets/50b119f7-e431-4e0d-9092-45eabeabe2d1)
![20260405_220821](https://github.com/user-attachments/assets/2fccef3e-279f-4917-9064-216888929fc3)
![20260406_023514](https://github.com/user-attachments/assets/b25851b3-97ab-46d4-b9f2-d52e5b31df07)
![20260406_024255](https://github.com/user-attachments/assets/3ee4b036-f3ba-4abd-8836-df570ec96803)
![20260406_030314](https://github.com/user-attachments/assets/e45d0c1e-584f-436e-b3f7-f897197f6905)
![20260406_030522](https://github.com/user-attachments/assets/769243d3-88b8-4539-a43f-10f00aafa48d)
![20260406_032609](https://github.com/user-attachments/assets/b0970371-259f-4d21-902b-bfc177ac433b)
![20260411_195727](https://github.com/user-attachments/assets/056cb66c-76a6-4236-9e15-eba4b11d5630)

 I did use ChatGPT for help, as I am not a studied electrical engineer, it was a painful journey but it works after all.
Removing the JC makes the device more independent and flashing is uninterrupted, but also takes away the pre-installed Ultra-Low-Power feature. But as we can see, the device is not lost even without the unit. A much easier workaround to run this device without the JC is to add mains power and keep tasmota always-on. For this, no further hardware-hack is necessary.
If someone here needs help with getting this sensor working on tasmota, I can try to help. I can also share my tasmota / home assistant setup. Cheers!

**Comment by yvoronkin** [comment]
If anyone is still interested in this hardware. I continued to reverse the board and at the same time assembled the configuration for tasmota. You can find it here [yvoronkin/docker-tasmota-for-tl-s157](https://github.com/yvoronkin/docker-tasmota-for-tl-s157), check the releases

Functionality:

- After connecting to MQTT, it sends the sensor status "Door closed" if there is a magnet and "Door open" if there is no magnet. 
- 15 seconds after waking up, it turns off its power with a command on js2303.
- If mqtt connection is not possible, power will be disconnected after 60 seconds
- Hold the reset button for 5 seconds - config will be reset

How to flash:
- Solder to uart pins
- On the R2 bridge (pulap for CHIP_EN), remove the solder blob after flashing, otherwise sleep will not work
- Pull R7 to ground during boot to enter flashing mode
- Flash the device `./esptool -p /dev/ttyUSB0 write_flash -fm dout 0x0 ./tasmota.bin`

Attention: Do not leave the solder blob on R2, otherwise sleep will not work
